### ISSUE TITLE
Switch to System.Numerics for relevant types

### ISSUE THREAD

**Issue by TinkerWorX** [MonoGame/MonoGame#7204]
[System.Numerics](https://docs.microsoft.com/en-us/dotnet/api/system.numerics?view=netcore-3.1) provides hardware accelerated Vector and Matrix types.

Current Candidates
- [ ] Vector2
- [ ] Vector3
- [ ] Vector4
- [ ] Matrix3x2
- [ ] Matrix4x4

What are the main challenges involved in switching to these?
1. Breaking namespace change.
1.1 We can provide overloads for the old types that will be marked as deprecated for a few versions.
2. Matrix4x4 and Matrix are the same type, but different names, so not straight forward.

**Comment by tomspilman** [comment]
The biggest concern we've had about this in the past is that `System.Numerics` passes big types like `Matrix4x4` by value and not by reference.  This is an optimization they didn't take from the XNA library and instead are depending on the C# compiler to do the right thing in these cases.

**Comment by harry-cpp** [comment]
> The biggest concern we've had about this in the past is that System.Numerics passes big types like Matrix4x4 by value and not by reference. 

We are already doing that, aren't we?

**Comment by perpetualKid** [comment]
just run a quick BenchMark.Net test using XNA Matrix Multiply both by val and by ref, System.Numerics Matrix 4x4 (which is by val), and handcrafted/optimized version for XNA Matrix multiply both by val, and using the new *in* modifier. 

RunXnaMultiplyVal `Matrix.Multiply(m1, m2)`
RunXnaMultiplyRef `Matrix.Multiply(ref m1, ref m2, out Matrix result)`
RunCustomMultiplyByVal `MatrixExtension.MatrixMultiply(m1, m2, out Matrix result)`
RunCustomMultiplyByIn `MatrixExtension.MatrixMultiplyByIn(in m1, in m2, out Matrix result)`
RunSystemMultiply `Matrix4x4.Multiply(m1, m2)`

on .NET Core 3.1, System.Numerics is clearly fastest even though passed by val

|                         Method |       Mean |     Error |    StdDev |
|------------------------------- |-----------:|----------:|----------:|
|               RunXnaMultiplyVal |   4.806 us | 0.0445 us | 0.0416 us |
|               RunXnaMultiplyRef |   4.071 us | 0.0290 us | 0.0271 us |
|     RunCustomMultiplyByVal |   3.613 us | 0.0722 us | 0.0675 us |
|      RunCustomMultiplyByIn |   3.343 us | 0.0300 us | 0.0281 us |
|              RunSystemMultiply |   2.305 us | 0.0234 us | 0.0219 us |

in .NET 4.8, there seems no benefit of System.Numerics

|                         Method |       Mean |      Error |     StdDev |
|------------------------------- |-----------:|-----------:|-----------:|
|               RunXnaMultiplyVal |   5.298 us |  0.1045 us |  0.1244 us |
|               RunXnaMultiplyRef |   4.358 us |  0.0863 us |  0.1060 us |
|     RunCustomMultiplyByVal |   3.957 us |  0.0926 us |  0.1622 us |
|       RunCustomMultiplyByIn |   3.530 us |  0.0683 us |  0.1084 us |
|              RunSystemMultiply |   5.575 us |  0.0503 us |  0.0446 us |

**Comment by mrhelmut** [comment]
This is a really open question: does anyone have a real world use case where using ```System.Numerics``` shows a noticeable difference?

On full games, these operations are likely invisible in a profiler, so just to make sure that we're not chasing rainbows.

Just a side note, but likely not relevant to the target use case (which I assume to be processing tools over games): these optimizations will not be used on consoles.

**Comment by bartwe** [comment]
Could you post source for the benchmarks @perpetualKid ?

**Comment by perpetualKid** [comment]
here is the code [ConsoleApp1.zip](https://github.com/MonoGame/MonoGame/files/4847556/ConsoleApp1.zip)

**Comment by perpetualKid** [comment]
so just for the fun of it did I do this switch to System.Numerics for relevant types: Matrix(4x4), Plane, Quaternation and Vector(2-4). Code convertion is pretty much straight forward, much simpler than first anticipated. Beside of removing the MG Types, adapt System.Numerics usings, and change Matrix to Matrix4x4, there are only minor code changes required. 
For those interested in the source, please see this fork https://github.com/perpetualKid/MonoGame/tree/System.Numerics

I've converted/tested WindowsDX and DesktopGL only, as well the Windows tools. 
Few observations:
- the types are mostly compatible and interchangeable. A very few missing methods (like `Vector*.Round|Ceiling|Floor|Hermite|CatmullRom`) needed to be addressed and are implemented as extension methods here https://github.com/perpetualKid/MonoGame/blob/System.Numerics/MonoGame.Framework/MatrixExtension.cs
- some other functionality missing in System.Numeric types are constants i.e. Vector/Matrix.Up|Down|Forward|Backward etc. For the few places used I just implemented those inline (i.e. `Microsoft.Xna.Framework.Vector3.Down` could be expressed as `System.Numerics.Vector3.UnitY * -1`)
- once done, some unit tests are failing. This is due to prebuild test assets like Tests\Assets\Fonts\Default.xnb which I couldn't find/figure the souce to rebuild. The tests in question fail as they try to load types references from Xna namespace. To fix this would need to rebuild the assets, and if different sources are used (such as different font), also change the reference images used for comparison. 
- I had upgraded the target framework for .NET from 4.5.2 to 4.8. Maybe another version in between (4.6.2. or 4.7.1) may have worked, but some System.Numeric types or methods are not available in 4.5.2. As side effect this would allow to use C#7.3 syntax (`ref readonly, in`) 
- Also the tools projects where upgraded to .Net Standard 2.1. Can't remember the exact reason, I think also some System.Numerics implementation was missing in 2.0 still.
- Couldn't find a way to attach Vector*.TypeConverter (from MonoGame/MonoGame.Framework/Design) to System.Numerics Vectors. However, I'm unsure of any usage, so dropped for now.

Overall, took me ~2 days of efforts, but this is PoC only, and with the constraints mentioned above, an implementation would only be feasible by leaving some compatiblity behind (namespace/name changes, framework support). Also, this would require testing for non-Windows platforms.

Now to get to a real-world scenario test :)
I've used the Monogame-version https://github.com/perpetualKid/ORTS-MG/tree/Core.MG.Numerics of [OpenRails](http://openrails.org/) train simulator (official OR is still using XNA, although there are efforts to move to MG also).
Again here, moving to System.Numerics and the above MG-version, adoption is straight forward. Importing namespaces, renaming Matrix, and recompile some effect shaders. Rebuilding, and - just works :) Both an .NET 4.8 and .NET Core 3.1 version are working fine.

For the impact: Not a reliable performance test, but boths version show a slight improvement of about 5-10% when comparing in-game FPS. Running through certain heavy scenery in the game which typically gets me barely to 20fps, now averages between 21 and 22 **WOW** The .Net Core version is bit slower (also from before) but shows similar improvement.

Conclusion (based on this very limited testing): There seems to be some (noticable) impact improving performance. While the changes (at least as required for this limited scenario) are managable, they come at a cost of breaking backwards compatibility. Once MG team is ready to do so, it's certainly worth adapting System.Numerics, reducing implementation surface with custom/own types over (.NET) framework-provided. The types implemented in MG are fairly mature and need low maintenance, and from MG perspective there is no immediate benefit of changing. If thinking of other changes (i.e. implement these types as readonly structs instead (?), this would be breaking change as well, so System.Numerics might be the more favorable way.

Also, for the tested scenario, while there seems some improvement, this does not justify to maintain a custom version of MG just to benefit. If MG though decides to go forward with this change, I'm happy to support efforts.

**Comment by mrhelmut** [comment]
Thank you for trying it on a full 3D project.

IMO if this requires breaking changes, we should not schedule it for MonoGame 3.x/4.x (which is intended to be XNA-compatible with some sort of LTS) but more likely something to think about when undergoing refactors/redesigns for what would be a future MonoGame 5.x (which is expected to be a departure from the XNA compatibility).

Another concern here would be to upgrade the target framework/C# language version. These are kept low because consoles can't handle language features which rely on .Net 4.6+ assemblies (but this is likely to be lifted at some point).

**Comment by vpenades** [comment]
I'm my case, I don't mind if Monogame keeps using it's own math types for backwards compatibility, _as long as it introduceds intrinsic conversions from/to System.Numerics types_ , because I'm pretty much fed up of writing ToXNA(); and ToSystem(); extensions for every new project I write using monogame.

```c#
public static implicit operator Microsoft.Xna.Framework.Vector3(System.Numerics.Vector3 v) { return ... };
```

This would allow developers to begin using BCL types right now, _without breaking compatibility_.

In fact, the replacement process can be done on two steps over a long period of time (1, 2 years):

1- As soon as possible, and during the rest of the life cycle of 3.x/4.x , introduce the intrinsic conversions and encourage developers to use BCL types instead of XNA types.

2- After the "adaptation period", replace the XNA types by the BCL types. Hopefully at this point enough projects will be already using the numerics types.

**Comment by Xeeynamo** [comment]
Hi, is there a pull request or any news to switch to System.Numerics? Other than performance I am also concerned about decouplation between my MonoGame.Framework front-end and the infrastructure libraries that are relying on System.Numerics.

**Comment by vpenades** [comment]
@Xeeynamo there's not been any development for a while.... I think there's a PR to move the pure math types to a separate package, but no progress on making these types compatible with System.Numerics.

The biggest roadblock is making Monogame.Framework to depend on System.Numerics.Vectors package. Although this seems a fairly simple dependency, Monogame developers have a lot of aversion to make the framework to depend on other libraries.

So first things first: Should monogame.framework depend on System.Numerics.Vectors?   thoughs? @tomspilman @harry-cpp @mrhelmut 

Afterwards, it would be easy to include the aforementioned implicit conversions, which would allow developers to use either Xna or numerics types seamlessly.

Sure, there might be some conversion overhead, but there's already conversion overhead for those of us who already need to interop between XNA and Numerics. So this is all about making developer's life easier.

**Comment by mrhelmut** [comment]
I would be curious about a real world benchmark in regard to moving to System.Numerics.

Other than that, as long as we have a replacement for runtimes where this isn't supported, we should be fine.

**Comment by vpenades** [comment]
I've just created #7417 PR to showcase an initial, and very basic System.Numerics.Vectors support. Let me know what you think.

**Comment by Tape-Worm** [comment]
> This is a really open question: does anyone have a real world use case where using `System.Numerics` shows a noticeable difference?
> 
> On full games, these operations are likely invisible in a profiler, so just to make sure that we're not chasing rainbows.
> 
> Just a side note, but likely not relevant to the target use case (which I assume to be processing tools over games): these optimizations will not be used on consoles.

This is not directly related to Monogame, but since someone asked for real world benchmarks, I figured I'd pass along what I discovered today. And no, this is not a "benchmark" per-se, but it's a pretty real world usage (for me at least).  

So, for the last couple of weeks I've been updating my graphics API, [Gorgon](https://github.com/Tape-Worm/Gorgon), to use System.Numerics instead of SharpDX's vector/matrix types... I got some positive feedback on matrix operations in a benchmark I did between SharpDX and System.Numerics (a bit of a speed boost for System.Numerics, not huge - can't recall the totals right now).

Anyway, today I got enough of it moved over to System.Numerics.  I ran my little benchmark example app which fills the window with bouncing ball sprites, 131072 of them. The balls move, scale, and change their opacity over time. 

* Before switching (SharpDX's vectors/matrices): 56-60 FPS (~18-16 mesc delta).  
* After switching (System.Numerics): 45-47 FPS (~21-23 msec delta).  
* After reverting back to the SharpDX types: 56-60 FPS again.

Just for reference, I'm using .NET Framework 4.7.2.  It's quite possible .NET 5/Core 3.1 would yield better results with System.Numerics, but until Microsoft gets their act together regarding UI, I can't switch to confirm this.

I'm guessing the difference is due to the fact that I can pass my parameters by reference in the SharpDX types, but clearly not in the System.Numerics. The SIMD doesn't seem to balance it out.

Anyway, I hope this helps.

**Comment by vpenades** [comment]
@Tape-Worm System.Numerics performance is highly dependant on the target framework AND the compiler you use. System.Numerics are SIMD accelerated only on a number of scenarios, one of those is, as you said, targeting Net Core 3.1 and above.

Also, a lot of methods are agressively inlined, which might produce misleading reasults in a number of scenarios.

For example: let's say you have a NetStandard 2.0 library using System.Numerics.... since NetStandard 2.0 doesn't have native support for System.Numerics, the methods that are agressively inlined within your library will use the regular code fo inlining. If you upgrade your library to NetStandard 2.1 (or NetCore) which _does_ support SIMD, the methods will be inlined using the SIMD code path, which is much faster.

So, if you write a library targeting netstandard2.0, and you consume it on a NetCore3.1 project, the methods in the library will use the slow code path because they've been inlined.

That's why, nuget packages that use system.Numerics maths agressively, are recomended to target netstandard 2.0 for compatibility, AND netstandard 2.1 (or netcore3.1) for performance.

But to me the whole point of using System.Numerics is not performance, but interoperability between libraries.

SharpDX types might be fast when you do a benchmark. But on a real project where multiple libraries need to act together, and you need to pass vectors and matrices between them, if the vector and matrix types are specific to each library, you need to convert between types. In other words, what you win in performance, you loose in type conversions.

Talking on benchmarks, I use to run tests in both Net4.7.1 and NetCoreApp3.1 .... and I use to get twice performance on the latter.

**Comment by Jjagg** [comment]
I agree this change is more about interop than performance. But as far as performance goes, on .NET Core/.NET 5+ perf of `System.Numerics` is better than what we have now in MG due to hardware acceleration.

> until Microsoft gets their act together regarding UI, I can't switch to confirm this

You can use .NET Core 3.1+ or .NET 5 with WPF/WinForms, if that's what you mean.

> I'm guessing the difference is due to the fact that I can pass my parameters by reference in the SharpDX types, but clearly not in the System.Numerics.

`ref` overloads were rejected in the .NET runtime because they did not impact performance (except for `Matrix4x4`):
https://github.com/dotnet/corefx/pull/25388#issuecomment-348360835

The JIT is very good at optimizing this stuff, and it's been improved a lot in .NET Core. Generally I would not recommend anyone ship a game with .NET Framework anymore.

**Comment by Tape-Worm** [comment]
> You can use .NET Core 3.1+ or .NET 5 with WPF/WinForms, if that's what you mean.

I'm aware of this, but the designer for WinForms in its current unfinished state is really not useful for me. I have a rather large application that uses WinForms and it's using 3rd party components that don't really work well under .NET Core/5 win forms. And converting it to WPF is a huge time investment, and the workaround of having a framework project for designing doesn't work for me. 

Also, in general, WPF has some rather ugly limitations when it comes to interop with Direct3D.

>  Generally I would not recommend anyone ship a game with .NET Framework anymore (it is also a huge pain for Mac/Linux).

Well, I don't really care about Mac/Linux support. I don't use either OS in my day to day, I'm perfectly happy being Windows only. And as stated, I'm unable to switch to .NET 5/Core, so what am I to do? Give up until they finish their work on WinForms?

Anyway, this is off topic.  I just wanted to share my findings with the Monogame devs. And while people are all drooling over .NET Core/5 (make no mistake, I am too), it's still not a reality for some of us, and our performance _will_ be impacted by switching to System.Numerics.

**Comment by Tape-Worm** [comment]
> @Tape-Worm System.Numerics performance is highly dependant on the target framework AND the compiler you use. System.Numerics are SIMD accelerated only on a number of scenarios, one of those is, as you said, targeting Net Core 3.1 and above.
> 

That's what I figured.  Hopefully one day I can join the rest of the world in .NET 5 or 6 or whatever land.

> Also, a lot of methods are agressively inlined, which might produce misleading reasults in a number of scenarios.
> 
> For example: let's say you have a NetStandard 2.0 library using System.Numerics.... since NetStandard 2.0 doesn't have native support for System.Numerics, the methods that are agressively inlined within your library will use the regular code fo inlining. If you upgrade your library to NetStandard 2.1 (or NetCore) which _does_ support SIMD, the methods will be inlined using the SIMD code path, which is much faster.
> 

Can't do it. I've tried 3 times now, and the major factor is the UI stuff (the unfinished forms designer just stops me cold).

> SharpDX types might be fast when you do a benchmark. But on a real project where multiple libraries need to act together, and you need to pass vectors and matrices between them, if the vector and matrix types are specific to each library, you need to convert between types. In other words, what you win in performance, you loose in type conversions.
> 

Wouldn't this still be the case if the libraries you're using don't use System.Numerics?

---

As I said in my other reply, I just wanted to post my results when it came to this stuff. There are probably many people using WinForms + Monogame that are in the same boat as I am, and it looks like their performance may possibly be affected by a jump to System.Numerics across the board.

**Comment by vpenades** [comment]
> Wouldn't this still be the case if the libraries you're using don't use System.Numerics?

Yes, at a cost. And it's been the case for a long time with a lot of libraries using their own types. You can still find important libraries that use their own implementations, WPF 3D and MathNet Numerics come to mind, with many more around. This is a consequence of System.Numerics being introduced so late into the Net library family.

Worst case scenario is using SharpDX vector libraries to base your math code on them, just to see SharpDX to be abandoned, making your code obsolete with them. That's why using BCL types is so important.

This doesn't mean that, _now we have BCL types_, and it is expectable that microsoft will optimize them in ways we can't, that should be a convergence of many libraries to adopt and use System.Numerics.

For example, ImageSharp, the new Bepuphysics, or my SharpGltf, use system.numerics, so these libraries can already interoperate seamlessly.

**Comment by vpenades** [comment]
@Tape-Worm Off topic: in our case we're not tied to WinForms, but we use WPF in some of our software. In our case we do projects for the medical sector, and they require us to keep supporting our software on Windows 7... and if it seems they're outdated, we've found many hospitals and madical centres still use Windows XP... so we can't expect our users using Win7 to upgrade anytime soon.

**Comment by Jjagg** [comment]
@Tape-Worm Could you share your benchmark project?

**Comment by mrhelmut** [comment]
Just to say that, with System.Numerics being a total no go on consoles (even the nuget backcompat), we'll be forced to provide our own non-intrinsics implementation of it, so users could simply rebuild MonoGame with the custom implem instead of the official support fairly easily (we could do that with two separate target files and a switch).

Whenever there's a PR for this, I'd be glad to also test this against our games (which are NET5).

**Comment by vpenades** [comment]
@mrhelmut my view one System.Numerics has always been to lay a roadmap with very small steps: being the first step to keep the monogame types, but allow intrinsic conversions. And only when Consoles are ready to go, either because an upgrade of BRUTE allows it, or because we switch to AOT compilation with Net6, only at that point it could be possible to completely replace the types, maybe for MonoGame 4.x

In the meantime, I've seen microsoft use the obscure _type forwarding_ that allows redirecting the use of a given type from one assembly to another, bypassing the default references. So maybe on consoles it could be possible to have a custom implementation called Monogame.Numerics, and use type forwarding, so whenever a System.Numerics usage is detected, it's redirected to the MonoGame.Numerics assembly.

https://docs.microsoft.com/en-us/dotnet/standard/assembly/type-forwarding

This has the advantage that third party projects using System.Numerics like the new BepuPhysics could run on consoles.

**Comment by Tape-Worm** [comment]
@Jjagg Sorry, I had removed the code in favor of SharpDX.  I restored it to Sys.Numerics tonight though, and did a .NET 5 conversion.  I've attached a zip file with my app (contains .NET 4.7.2 and .NET 5 versions).  Both versions are using System.Numerics.  I'd like to know what your performance is like (and of course the hardware on which you're running).

For me, I'm running it on an Intel 10980 CPU with an Nvidia 3090 GPU.

For my tests, I pump up the number of balls to 131,072.  In .NET 4.7.2, I'm getting 48-50 FPS (CPU bound), in .NET 5, it's abysmal at ~35 FPS. Seriously bad, and I was totally expecting a boost when going to 5 due to SIMD and whatnot (maybe I'm not using it correctly?)

You'll require a GPU that supports Direct3D 11.2 at minimum.

[BallsPerTest.zip](https://github.com/MonoGame/MonoGame/files/5977136/BallsPerTest.zip)

**Comment by roeyskoe** [comment]
@Tape-Worm There is something weird going on. 
On my i7-5820k and GTX 970 i get 30-35 fps on .NET 5.0 and 27-29 fps on .NET 4.7.2 when having exactly 150k balls.
Yes the difference is not big, but I would expect getting way worse results than you, since you have a lot more horsepower than I do.

On both tests my CPU usage stays very low, around 13%, highest core at max 50%. GPU is at around 70% on net472, and 80% on net5.

**Comment by vpenades** [comment]
@Tape-Worm I did download your test but did not run it, but I understand your test includes graphics rendering. I'm not sure what you're testing, but the difference in framerate between net472 and net5 might be by a lot of reasons, not only because using one math library or another.

A proper performance/benchmark test must use only the library to be tested alone.

**Comment by mrhelmut** [comment]
A proper test would be to benchmark a full game with and without intrinsics, both on the same runtime. If we're comparing older runtimes with the latest ones, we can't really assess if anything is related to intrinsics being used or the inherent optimizations/differences of runtimes. We also have to make sure that both sides run on the 64bit runtime.

I'm not quite into synthetic benchmarks or apps built on purpose to stress test a precise functionality because they don't consider the average cases and are not optimized in the same way by the JIT compiler (which isn't quite magic in that regard).

@perpetualKid Is there any chance that you could rebase your branch on the latest ```develop```? I would be keen to give it a test spin on our projects.

**Comment by Tape-Worm** [comment]
@roeyskoe What the???  Really?  Oh man, something's definitely not right on my end then.  Time to crack open the profiler I guess.

**Comment by Tape-Worm** [comment]
The purpose was to see how the numerics stack up against each other, I was just so surprised by the poor performance in NET 5 I included it, but I also should have included the original SharpDX-math versions as well.  I apologize for any confusion this may have created.

I've updated the test application on my previous post to include the apps using the SharpDX math versions.

My results for these were quite surprising:
.NET 4.7.2, SharpDX Math -> 56-58 FPS
.NET 4.7.2 System Numerics -> 49-51 FPS
.NET 5 SharpDX Math -> 50-52 FPS
.NET 5 System Numerics -> 38-40 FPS

The only code that's been changed between the versions is the math functionality, that's it. All other code is identical.

One thing that occurred to me is that I'm using lots of Vector2/Vector3 values, and it could be that if I were using Vector4 values primarily, I might see an improvement (I've read that Vector4 sees gains from SIMD, and it makes sense to me that it would). With that said, doing that doesn't fit with my API surface (I'm in 2D, why do I need Z and W?).

I'm aware this is not the same as comparing with BenchmarkDotNet or the like, but I wanted to see how these math APIs would compare in an environment where they were being used for more real world-like things (e.g. rendering lots of objects to the screen), because, after all, this is what it's going to be used for in my case.  And the same would apply to those using Monogame as well.

**Comment by initram** [comment]
Please make sure that you are running it in a way that supports SIMD, try to test whether `Vector.IsHardwareAccelerated` is true: https://docs.microsoft.com/en-us/dotnet/api/system.numerics.vector.ishardwareaccelerated?view=net-5.0#System_Numerics_Vector_IsHardwareAccelerated

Vector2 should also benefit from SIMD, at least according to this: https://docs.microsoft.com/en-us/dotnet/standard/simd

**Comment by Tape-Worm** [comment]
@initram I verified that IsHardwareAccelerated is true.

**Comment by Zintom** [comment]
Is there any update for this? I'm all for a switch to the System.Numerics namespace.

**Comment by vpenades** [comment]
@Zintom I feel like it's not gonna happen anytime soon; The best I've been able to do is to introduce implicit conversions between system.numerics and monogame types (MonoGame 3.8.1-dev), so you can already call monogame methods as if the types already are system.Numerics.

It's not optimal, but at least it's one step forward in the right direction. If enough people uses monogame in this way, there will be enough traction to switch to system.numerics, probably at monogame 4 or later.

**Comment by Zintom** [comment]
> enough traction to switch to system.numerics, probably at monogame 4 or later.

Couldn't we just do the following: Make the whole framework use System.Numerics and provide implicit operators(both ways) on the Framework.Vector2 type (like you have), that way any old code would still work and any calls to the framework would implicitly convert from the passed Framework.Vector2 into the System.Numerics equivalent, granted this might slightly slow down the performance of user code that didn't switch to the new System.Numerics namespace as every method call that accepted a Vector2 would be creating a new struct on every invocation.

On another note, why the heck did it take the .NET team 18 years to bake in support for Vector types smh 🙃🤪. Instead they're like ouu lets implement Top Level Statements in C# 9 😄

**Comment by vpenades** [comment]
> use System.Numerics and provide implicit operators(both ways)

In my experience, it's not wise to implement two-way implicit converters because under a number of scenarios, it can mislead the compiler and you end up with compiler errors due to unresolvable ambiguities.

I agree that System.Numerics.Vectors where introduced way too late into .Net, but we have to live with that.

Anyway, simply adding the System.Numeric.Vectors nuget package dependency has caused a number of problems with the MonoGame console backends, so for now, the implicit conversions is the best we can ask for now.

I don't think there'll be any further development until the console issues are resolved and the staff feels comfortable moving forward.

At which point, the next step world be to replace the types in the method calls of all MonoGame APIs, and reversing the implicit converters in the framework, so old code using the old framework vector types world be implicitly converted to system numerics on method calls... But the leads have stated that such breaking change would not happen in MonoGame 3.8.x... So I guess it's a change for MonoGame 4

**Comment by mrhelmut** [comment]
We could switch to ```System.Numerics.Vectors``` if our local implementation for consoles (https://github.com/MonoGame/MonoGame/tree/develop/MonoGame.Framework/Utilities/System.Numerics.Vectors) is expanded to be feature complete (right now it's just a dummy implementation).
We could also provide a simple define in the csproj to enable/disable SIMD support with this implementation (SIMD is not always desirable, if code is not specifically designed for vectorization there's a very high chance that the SIMD instructions are bottlenecked and have a negative impact on performances, they're not magic and I don't think we should expect benefits from them for 90% of devs). My suggestion would be to have SIMD disabled by default using the local implementation, and allowing whoever _needs_ SIMD to switch to intrinsics.

Anyway, these changes are breaking changes, which is something we should expect by MonoGame 5.x (MonoGame 4.x is expected to be the LTS "XNA 4 compatible" branch, with MG 5 departing from it and breaking some stuff).

**Comment by frankhaugen** [comment]
When .net runtime have types that cause monogame to have duplicates, the .net types should be used.

 I'd like to see Monogame move to an in-step versioning with the runtime and depend directly on it. It would break people, but some regular breakage is good, as it's unsustainable to expect no breakages and get regular updates at the same time

**Comment by Zintom** [comment]
> It would break people, but some regular breakage is good, as it's unsustainable to expect no breakages and get regular updates at the same time

I totally agree with this sentiment. I think the main devs are waiting till 4.0 to do any large changes. Moving from the Monogame numeric types to System.Numerics should be a 4.0 feature, along with moving out of the Microsoft.XNA namespace. Yes System.Numerics might not perform as well on .NET Framework or may need a nuget package to work, but in all honesty .NET Core (or new .NET or whatever you want to call it) is the future, I don't know anyone starting projects for .NET Framework.

**Comment by vpenades** [comment]
@frankhaugen Monogame vector and matrix types predates those of System.Numerics.Vectors, in fact, System.Numerics.Vectors types were lifted from XNA by microsoft in the first place.

But now that .Net has in-build types, we all agreed monogame show move towards a replacement.

A while ago I submitted a PR that allows using System.Numerics.Vectors in monogame using implicit conversions, which is not ideal, but it provides some level of interoperability without breaking compatibility. This PR was merged, but it hasn't reached the nuget packages yet.

In a far, far, future, the idea is to reverse the dependency, so monogame would use System.Numerics.Vectors, but also without breaking compatibility for most cases.

**Comment by damian-666** [comment]
Might consider waiting on this until this way is legal , might be relevant: 

 https://pvs-studio.com/en/blog/posts/csharp/0878/

https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/static-abstract-interface-methods
  btw  in net 6 passing  input by ref isnt worth the mess, it's already optimized.       for most engines that have put in alot of SIMD work that ive looked at, the bottlenecks  require custom array processing using much larger vectors,  than 2 or 3 d vectors or 4x4 matrices.. the processing done in monogame itself,   is so minimal i dont think we will to see any substantial, as in order of magnitude gains, but be nice not to have to copy types..

**Comment by vpenades** [comment]
@damian-666 thing is, many things in monogame are optimized for the "beast" compiler so monogame can be used in consoles. I don't know the current status of this compiler since it's only offered to private developers working on consoles. But last time I was aware of it, it was on par with what Net Framwork 4.5 was able to do.

That's the main reasons because system.numerics.vectors was not adopted sooner, and it is unlikely it will ever be. Same thing goes for other newer net sugar, like Span<T>

**Comment by damian-666** [comment]
I get it.. I'm just so glad now I can use NetStandard to
compile.my.core.stuff but then when I use a  thin net6 console  client as
an exe, then  my framerste flies and I regret some of the optimization i
did early on that mattered. then bit now just a bunch of junk code..
 I appreciate.rhe fact they don't use the latest stuff or accept many
changes.  I don't mind  to have copy stuff or make conversions and frankly
I think I might wait and see on  SIMD because I feel it could be
generalized and when I see intrinsics it looks like a horror show to code
one even resorted to managed cplus plus for a fluid simulation.array.
..in image sharp and bepu I think they got a good grip on it but it's mind
boggling stuff  with partitioning to vector 128  and cache awareness,  I
would hope that the runtime could eventually do this  for us like task
parallel did.. anyways long live monogame!

**Comment by mrhelmut** [comment]
Yeah, MonoGame is very much bound to the console compiler which still needs apps to target .NET 4.5 (and not just targeting it, it has to use the 4.5 BCL and compiler ; targeting net45 with net6 won't do). This isn't something that is expected to change in the foreseeable future. Since console developers are the main audience of MonoGame and the maintainers have no bandwidth for more, there is little to no room for having MonoGame to be a shiny innovative library. It is more important for MonoGame to provide an LTS support than being up-to-date with the latest .NET innovations.

If you need a .NET graphics library for heavy processing that aren't games, MonoGame is likely not that library.

**Comment by damian-666** [comment]
i just did some benchmark tests with a version of Farseer
https://github.com/ebatianoSoftware/Farseer-Physics-Standard, starting
things like decomposition that can take long, im not seeing any improvement
, in fact its a tiny bit slower in numberics,  but the new c# 10 feature "
global using" System.Numerics,  for just the vector types instead of the
analog xna types... and  global using Farseer.Xna.Framework for the other
exe , is making this comparison test all possible without repeat code.
but ill do more tests and if i find anything that gives a decent boost ill
report back otherwise i dont see any performance boost at all so far.
 Anyways thought id mention the new global using is extremely helpful
otherwise id have no way to share the test code between the rigs..   Im
not suggesting using new c# features  in Monogame, just its a useful way in
ur own EXE code that i just discovered.

**Comment by vpenades** [comment]
@damian-666 I think everybody takes wrong the issue of system numerics.

The whole point of switching to system numerics was not about performance improvements... or at least not the main reason. The main reason has always been to have a common type that can interop with other libraries that don't depend on monogame.

It really doesn't matter if monogame's vectors and system vectors perform the same; the real performance loss happens when you need to continuosly convert between the two. My code is chocked full with .ToXNA() and .ToNumerics() extensions. This is not only a pain during development, it's what's causing the biggest performance loss regarding using math types.

**Comment by mrhelmut** [comment]
> @mrhelmut This is not the place of discussing the state of Monogame, but I'll just say that by focusing on consoles and disregarding the rest of non-console developers, monogame is loosing atractiveness, and this will end affecting the whole ecosystem, consoles too.

@vpenades  We're completely aware of this, and it is not a situation that we enjoy. I just wanted to be transparent on the state of the project. The admins have a lot of cool plans to move MonoGame forward (e.g. switching to Vulkan/DX12/Metal) and some great PR are pending (e.g. compute shader and shader conductor), but our bandwidth is limited so our attention is where support is actively requested. We lack the manpower to move innovations forward.

Ideally we would like to welcome new people and encourage contributing, and we've been working on a call to participate and an explanation of the possible future of the project, as well as discussing how we could leverage funds to help, but these discussions are moving slowly. I'm trying to take the lead but the time I can allocate is irregular.

**Comment by MrGrak** [comment]
> This is not the place of discussing the state of Monogame, but I'll just say that by focusing on consoles and disregarding the rest of non-console developers, monogame is loosing atractiveness, and this will end affecting the whole ecosystem, consoles too.
I don't usually comment here, but I felt compelled to, after reading this. There's some issues with this statement that I'd like to address/unpack.
1. A lot of developers are attracted to Monogame because of console support, plus being open source. Myself included. I see new developers joining the discord and asking about consoles every month. 
2. Do you have any data or metrics to back up your opinion that Monogame is loosing attractiveness due to this issue? 
3. The majority of games made in MG don't make use of numerics, nor do they have performance loss issues related to continuously converting between the two. 
4. "and this will end affecting the whole ecosystem, consoles too." - You're using fear to motivate change. I don't think you can predict the future or know what's going to happen to the ecosystem. I don't think you even have data to back up your opinion.
5. If you want to see specific changes, why don't you contribute those changes? Or work with contributors to get those changes into MG? You're essentially complaining that volunteers aren't working fast or hard enough on the changes you want. 

The future of Monogame is bright, and I trust in the volunteers and contributors to guide the ship as they have done in the past. I'm looking forward to their future plans, and keeping an eye on FNA too. Hope this post doesn't upset anyone.

**Comment by vpenades** [comment]
@mrhelmut I think the reason I've dropped the issue here is becase there's no place left to have these kinds of discussions, I've tried to ask a while ago at discord with no answer.

@MrGrak: I have reasoned answers for all your questions, but they're long, as I said I don't think this is the right place, and I don't know which is the appropiate channel.

**Comment by damian-666** [comment]
have some profound results here with net 6 on win64 for numerics.Vector2 on adding and normalizing.. w numberics Vector2, im seeing that it is apparently using SIMD and is 2.5x faster on a simple test vs xna vector2 from my old farseer copy.   Another successful test, 30-40% fastr, is its  use in concave polygon decomposition that  must do before filling w MG.    i'm   not pushing for any breaking changes to MG because i don't believe its will make such a profound effect at all for its methods,  or any at all,  but for complex physic or decomposition its clients should definitely use numerics and then copy the structs because, i got  the blazing 2.5x boost on vector2 normalize and add, even w copy.   I might bring it up in the 2D physics engines issues  its just there is no canonical often maintained branch now.  ..

  And I also found that  the cost of copy the struct of vector2,  is comparatively negligible.  . I did a copy to vec struct in addition to the add and normalize only on the numerics version test and its still 2.5x faster.    The issue i found with the complex code like decomposition and physics (outside of MG but commonly used by clients) is that there they predate numerics and use similar matrices and other custom types and not operating standard matrix and vector arrays so we dont see the 2x+ boosts we could expect.    first quick rough test i was just very curious.    But I def suggest using these types in your own complex code.   
  heres my  rought tests:  
https://github.com/damian-666/PhysicsBenchmarks

i don't do statistics because if  its not 50% or 2x faster i don't bother with optimizations im looking for orders of magnitude in unexploited potiential like SIMD.

i dont know the source fork of https://github.com/ebatianoSoftware/Farseer-Physics-Standard     Digging deeper, mabye these engines froks could be sped up, i tested TOI it didnt help that,  but i didnt test other aspects for now im waiting for net 6 on xamarin , got enough performance on pc.

**Comment by RoyAwesome** [comment]
Hi

Given that the most recent version of MonoGame only targets .net 6 and the performance gains of using the Numerics types is well established nowadays, is there any chance of movement on this issue?

**Comment by vpenades** [comment]
@RoyAwesome The latest version of MonoGame has _implicit conversions_ between Numerics and MonoGame types (Vector2,Vector3, Matrix, Quaternion, etc), so you can already use Numerics in your code transparently.

The MonoGame APIs still use the old math types, but given that backwards compatibility is a top priority, I don't think big changes will come anytime soon.

**Comment by mrhelmut** [comment]
> is well established nowadays

Synthetic benchmarks are irrelevant for games. We're still waiting on real world tests of a full game with and without Numerics. 

I've already explained it above, TL;DR: intrinsics are not magic, enabling them doesn't always yield positive results. The examples that we have of real full games so far show worse performances with intrinsics. There can be many reasons for that to happen, but the most common is that intrinsics is not what you think it is. It is not something that just swap an instruction for another faster one. It is something that batches multiple operations, and if you don't "pack" your vector operations in your code, you can pretty easily bottleneck the CPU pipeline because it's not batched properly (it will basically spin doing nothing).

Taking advantage of intrinsics requires that your code [is designed for it](https://devblogs.microsoft.com/dotnet/hardware-intrinsics-in-net-core/). If you place Numerics operations at random (like in a game code base or monogame itself), all you're going to do is feed the CPU with incomplete vectors, which means that the CPU will be spending more time than it should.

**Comment by damian-666** [comment]
This is the best way IMO..I look at Vector2 in the root of the source  of
monogame  and its  look s perfect for me now and the implicit conversion
very welcome. I see the the class its also now persistent  with
Datacontract, which is why i have my own Vector2 class based on the old
Farseer one ( I  use data contract serializer) to save levels and
collections of vector2 ..so  i will try and change to this class probably.
  also theres other pitfalls,  evidence 32 bit float simd is faster than on
64 bit machine ( obviously for floats you are wasting loading up those
lanes with unused data ) in net core discussion , its unsolved) .. so yea
its not always a win though a simple test of normalization + add random
vector.. is 2.5x faster on my machine which is .5 x faster than i dreamed
possible for a 2d vector.   but its your own code that your physics and
other math has the bottlenecks and then let the conversion do its thing
later... monogame isn't doing heavy complex maths on CPU , as far as i can
guess..  im my experience the copy costs nothing compared with complex
algorithms your game might be doing.

**Comment by orosbogdan** [comment]
> @mrhelmut the purpose of the implicit conversions is not (and never has been) to improve performance, but to make it easy to interop with code that does use Numerics (BepuPhysics, ImageSharp, etc), which is becoming more and more common.
> 
> Also, implicit conversions serve a purpose in the long run: if the long term plan is to eventually move from XNA math types to Numerics, those that program their game logic using Numerics will take now some performance loss due to the implicit conversion, but eventually, when (if) monogame ends supporting Numerics all the way down to the core, those who already use Numerics would see a huge performance improvement.
> 
> I am aware that game lifecycle is probably 1~2 years, so most games in development would do better by sticking to XNA types. But long term projects like ours, and maybe third party libraries that might want to stay around for years might consider using Numerics to avoid the high cost of changing types and breaking the API at some point in the future.
> 
> Regarding _intrinsics_ , it's true that regular operations might not show any performance improvement, but there's more to that: for example, XNA's math use to pass arguments by ref, whereas Numerics pass arguments by value, this might look like an issue, but in fact, Net6 has lots of optimization paths for Numerics that can really improve the performance (or so I've been told)... what's sure is that Microsoft is going to keep pushing performance on Numerics.

Seconding this, interation with other libraries, especially physics one would be so much cleaner.

**Comment by warquys** [comment]
I made a rough sketch of what will be possible to do via the new extension syntax.
This still causes a breaking change. But, the rewrite is minimal (only the Matrix type needs to be renamed to Matrix4x4).
https://github.com/warquys/MonoGame/tree/SystemNumeric
The extension syntax is from language version 12, still in preview. I will probably change it in the future. Currently, the CLI does not compile it well. I don't know if I have two installs of dotnet or if the preview language is only available on VS.

I tried to see if it was possible to have an unsafe cast to call the System.Numerics methods directly. It's possible, but the compiler can't figure out the pattern and causes more slowdowns than anything else. This would have allowed MonoGame to be updated on System.Numerics without having to make a breaking change.

This seems strange to me because I get good results when I cast Numeric Vector4s to Vector128s, and the same goes for Vector2s to Vector64s.
But not for XNA Vector to Numeric Vector.

I ended up rewriting the thing by overlapping the X, Y, Z fields on a Vector3. The performance is almost that of a Numeric Vector3. I say almost because most of the time when I run benchmarks, I get results slightly slower than Numeric vectors. However, for some reason I can't explain, I get times 10x faster than the Numeric system. Which seems impossible to me. I think the CPU notices a repetition and decides to make some obscure optimization.

On my computer, the Numeric namespace is 40x faster than current vectors.
https://github.com/warquys/MonoGame/tree/NumericOverlapping

However, I was unable to pass all the unit tests and only verified that it compiled under MonoGame.Framework.DesktopGL.
First issue: `Dot` under SIMD does not give the same result, as a writen compute `X * X + ...`. Some operations may possibly return values depending on the CPU architecture and the way in which the architecture manages rounding.

I was not able to disable the SIMD to test the performance without SIMD.

However, even though some may be tempted by this idea of overlapping,
this always causes a breaking change in the code syntax. The compiler will require all fields to be initialized when the following syntax is used:

```cs
void SomeMethod(out Vector2 vec)
{
    vec.X = 10;
    vec.Y = 20; // ERROR: does not initialize all fields (missing init of Numeric.Vector2 backing field).
}
```
I assume this is not a breaking change for already compiled code (I tested in .NET, IDK if that's the case for Framework).

My benchmarks are here:
https://github.com/warquys/BenchmarkRefactorNumeric
